### PR TITLE
Adding native support for strong naming assemblies.

### PR DESCRIPTION
**Pull request by moozzyk**

### CODE REVIEW

File: test/Microsoft.Framework.Runtime.Tests/ProjectFacts.cs [hunk 1]
```diff
@@ -373,7 +373,7 @@ public void CompilerOptionsAreSet()
         {
             var project = Project.GetProject(@"
 {
-    ""compilationOptions"": { ""allowUnsafe"": true, ""define"": [""X"", ""y""], ""platform"": ""x86"", ""warningsAsErrors"": true, ""optimize"": true }
+    ""compilationOptions"": { ""allowUnsafe"": true, ""define"": [""X"", ""y""], ""platform"": ""x86"", ""warningsAsErrors"": true, ""optimize"": true, ""strongNameKeyFile"" : ""c:\\keyfile.snk"", ""delaySign"" : true }
```
Comment (pranavkm) [inline test/Microsoft.Framework.Runtime.Tests/ProjectFacts.cs]: Can we get a couple of variants for the `delaySign` values ('1`, '-1`, `True`, `""`)?

**Comment by pranavkm** [comment]
Looks good. Would like to see more test cases before :shipit:

File: test/Microsoft.Framework.Runtime.Tests/ProjectFacts.cs (hunk 1 above)
Comment (moozzyk) [inline test/Microsoft.Framework.Runtime.Tests/ProjectFacts.cs]: Not sure if I get the purpose - wouldn't it be just like testing the JSON parser (which I assume is already tested)? We only allow `true/false` `bool` values in the project.json and I assume `rawOptions.ValueAsNullableBoolean("delaySign")` will break on everything else. I left `1`/`true` for environment variables to not break anyone who is already using it (e.g. NuGet).

**Comment by moozzyk** [comment]
I tested the changes manually on the dnx.clr.managed.dll (which is the target for this change)

File: test/Microsoft.Framework.Runtime.Tests/ProjectFacts.cs (hunk 1 above)
Comment (pranavkm) [inline test/Microsoft.Framework.Runtime.Tests/ProjectFacts.cs]: Nevermind - didn't realize the place you were string comparing `1`, `true` is the env variable. Ignore my comment

Comment (moozzyk) [inline test/Microsoft.Framework.Runtime.Tests/ProjectFacts.cs]: No worries :)

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs [hunk 2]
```diff
@@ -192,17 +193,21 @@ public class RoslynCompiler
...
-                    string.Equals(delaySignString, "true", StringComparison.OrdinalIgnoreCase) ||
-                    string.Equals(delaySignString, "1", StringComparison.OrdinalIgnoreCase));
+                var delaySign =
+                    delaySignString == null
+                        ? compilationContext.Compilation.Options.DelaySign
+                        : string.Equals(delaySignString, "true", StringComparison.OrdinalIgnoreCase) ||
+                          string.Equals(delaySignString, "1", StringComparison.OrdinalIgnoreCase);
 
-                var strongNameProvider = new DesktopStrongNameProvider();
+                var strongNameProvider = new DesktopStrongNameProvider(ImmutableArray.Create(Directory.GetCurrentDirectory()));
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: That's the wrong directory. Use the project directory of the compilation context

Comment (moozzyk) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Will do

Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: This entire strong naming code path also needs to be skipped on mono.

Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: You can fake sign on any platform if no key is provided and delay sign is true. We should also support that.

Comment (moozzyk) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: do you mean we should not throw the error if delay sign is true and the file key is not provided and just skip the code? Why would it be useful?

Comment (moozzyk) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Fake signing will be separate.

**Comment by moozzyk** [comment]
Updated with:
- changed `strongNameKeyFile` to `keyFile`
- fixed the issue with the snk file search location - using `compilationContext.Project.ProjectDirector` instead of `Directory.GetCurrentDirectory()`
- added a check to fail on Mono

**Comment by moozzyk** [comment]
:watch:

**Comment by davidfowl** [comment]
:shipit:

**Pull request merged**